File: eunji/0x08_stackApp/BOJ_4949.cpp
#include <iostream>
#include <deque>
using namespace std;

void check(deque<char> &S, char input, string &ans){
    if (input == '(' || input =='['){
        S.push_back(input);
    }
    else if (input == ')'){
        if (S.empty()) ans = "no";
        else if (S.back() == '(') {
            S.pop_back();
        }
        else ans = "no";
    }
    else if (input == ']'){
        if (S.empty()) ans = "no";
        else if (S.back() == '[') {
            S.pop_back();
        }
        else ans = "no";
    }
    else return;
}
int main(void){
    ios::sync_with_stdio(0);
    cin.tie(0);
    string input;
    string ans = "yes";
    deque<char> S;
    while (1){
        getline(cin, input);
        // cin >> input;
        if (input == ".") break;
        for (auto c : input){
            if (ans == "yes"){
                check(S, c, ans);
            }        
            if (c == '.') {
                if (S.empty() && ans == "yes") cout << "yes\n";
                else cout << "no\n";
                ans = "yes";
                S.clear();
            }
        }
    }
}

File: eunji/0x08_stackApp/BOJ_4949.py
from collections import deque
import sys

def check(input):
    if (input == '(' or input == '['):
        S.append(input)
    elif (input == ')'):
        if (len(S) == 0): return "no"
        elif (S[-1] == '('): S.pop(); return
        else: return "no"
    elif (input == ']'):
        if (len(S) == 0): return "no"
        elif (S[-1] == '['): S.pop(); return
        else: return "no"
    else: return
    
ans = "yes"
S = deque([])
while True:
    input = sys.stdin.readline()[:-1]
    if (input == "."): break
    for c in input:
        if (ans == "yes") and check(c):
            ans = check(c)
        if (c == "."):
            if (len(S) == 0 and ans == "yes"): 
                print("yes")
            else: print("no")
            ans = "yes"
            S = deque([])